default.nix

```diff
@@ -39,92 +45,125 @@ let
+  featuresFlag = if enabledFeatures == "" then "" else "--features " + enabledFeatures;
+  packageFlags = builtins.map (crate: "--package " + crate) (builtins.attrNames crates);
 
   # allows to skip files listed by .gitignore
   # otherwise `nix-build` copies everything, including the target directory
-  gitignoreSrc = nixpkgs.fetchFromGitHub {
-    owner = "hercules-ci";
-    repo = "gitignore.nix";
-    rev = "5b9e0ff9d3b551234b4f3eb3983744fa354b17f1";
-    sha256 = "o/BdVjNwcB6jOmzZjOH703BesSkkS5O7ej3xhyO8hAY=";
+  inherit (versions.gitignore) gitignoreFilter;
+  # we need to include the .git directory, since Substrate build scripts use git to retrieve commit hash of HEAD
+  gitFilter = src:
+    let
+      srcIgnored = gitignoreFilter src;
+    in
+      path: type:
+        builtins.baseNameOf path == ".git" || srcIgnored path type;
+  src = nixpkgs.lib.cleanSourceWith {
+    src = ./.;
+    filter = gitFilter ./.;
+    name = "aleph-source";
   };
-  inherit (import gitignoreSrc { inherit (nixpkgs) lib; }) gitignoreSource;
 in
-with nixpkgs; env.mkDerivation rec {
-  name = "aleph-node";
-  src = gitignoreSource ./.;
-
+with nixpkgs; naersk.buildPackage rec {
+  inherit name release src singleStep;
+  doCheck = runTests;
+  nativeBuildInputs = [
+    git
+    cacert
+    pkg-config
+    llvm.libclang
+  ];
   buildInputs = [
-    rustToolchain
-    llvm.clang
     openssl.dev
     protobuf
-    customRocksdb
-    pkg-config
-    cacert
-    git
-    findutils
-    patchelf
-  ];
-
+  ] ++ nixpkgs.lib.optional useCustomRocksDb customRocksdb;
+  cargoBuildOptions = opts:
+    packageFlags
+    ++ [featuresFlag]
+    ++ ["--locked" "--offline"]
+    ++ opts;
   shellHook = ''
-    export RUST_SRC_PATH="${rustToolchain}/lib/rustlib/src/rust/src"
-    export LIBCLANG_PATH="${llvm.libclang.lib}/lib"
-    export PROTOC="${protobuf}/bin/protoc"
-    export CFLAGS=" \
-        ${"-isystem ${llvm.libclang.lib}/lib/clang/${llvmVersionString}/include"} \
-        $CFLAGS
-    "
-    export CXXFLAGS+=" \
-        ${"-isystem ${llvm.libclang.lib}/lib/clang/${llvmVersionString}/include"} \
-        $CXXFLAGS
-    "
-    # From: https://github.com/NixOS/nixpkgs/blob/1fab95f5190d087e66a3502481e34e15d62090aa/pkgs/applications/networking/browsers/firefox/common.nix#L247-L253
-    # Set C flags for Rust's bindgen program. Unlike ordinary C
-    # compilation, bindgen does not invoke $CC directly. Instead it
-    # uses LLVM's libclang. To make sure all necessary flags are
-    # included we need to look in a few places.
-    export BINDGEN_EXTRA_CLANG_ARGS=" \
-        ${"-isystem ${llvm.libclang.lib}/lib/clang/${llvmVersionString}/include"} \
-        $BINDGEN_EXTRA_CLANG_ARGS
-    "
-    export ROCKSDB_LIB_DIR="${customRocksdb}/lib"
-    export ROCKSDB_STATIC=1
+    ${rocksDbShellHook}
+    export SUBSTRATE_CLI_GIT_COMMIT_HASH=${SUBSTRATE_CLI_GIT_COMMIT_HASH}
+    export RUSTFLAGS="${rustflags}"
+    export LIBCLANG_PATH=${LIBCLANG_PATH};
```

Comment:
Similar question to all this envs, who sets them in our current build process, and what are their values? Cargo does it?

Reply:
Mostly, they are not set by cargo, just some magic happens in some of the build.rs scripts which are trying to find them in some `standard` paths. Unfortunately, those paths are meaningless under nix-build.